# Umi.seq: Ultra-deep targeted sequencing of a narrow colorectal cancer-relevant subset of the genome to a unique coverage of 5.000-30.000x

I developed a tumor-agnostic approach for the detection of colorectal adenomas and cancers (CRC) based on the DNA fragment length distribution on- and off-panel. It is based on a novel methodology for the processing of plasma samples collected from CRC patients postoperatively and collect serial samples from the IMPROVE clinical trials ([NCT03637686](https://clinicaltrials.gov/ct2/show/NCT03637686), [NCT03748680](https://clinicaltrials.gov/ct2/show/NCT03748680), [NCT04084249](https://clinicaltrials.gov/ct2/show/NCT04084249)). For tumor-informed analysis, whole-exome sequencing data are generated from resected tumor tissues and matching peripheral blood mononuclear cells before surgery. After surgery, a targeted panel of 12,500 bp based on a 12-gene panel and 90 known CRC mutations is sequenced and labeled using unique molecular identifiers. The analysis matrices are of size five (A, C, G, T, and - for deletion) by 18,094, which accounts for a five basepairs distribution around each mutation. My fragmentomics analysis allowed me to identify a dysregulation in the antioxidant program as a possible biomarker of sub-clinical and early-stage CRC, as 86% of Stage I CRC tumors do not shed DNA. Here, I present a technical evaluation of the sensitivity and specificity of the CRC panel to correctly detect variants of the CRC mutations based on different amounts of input DNA and read coverage. 
